src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+
+  // TODO(anybody): Is this check actually necessary, or is this always true for correlated parameters?
+  // Check that all correlated parameters expressions are column expressions of the left sub-tree. Otherwise, we won't
+  // be able to use them in join predicates. Also build up a map from parameter ids to their respective expressions.
+  std::map<ParameterID, std::shared_ptr<AbstractExpression>> parameter_mapping;
+  for (size_t parameter_idx = 0; parameter_idx < subquery_expression->parameter_count(); ++parameter_idx) {
+    const auto& parameter_expression = subquery_expression->parameter_expression(parameter_idx);
+    if (!left_tree_root->find_column_id(*parameter_expression)) {
+      _apply_to_inputs(node);
+      return;
+    }
+
+    parameter_mapping.emplace(subquery_expression->parameter_ids[parameter_idx], parameter_expression);
+  }
+
+  // Find usages of correlated parameters and try to turn them into join predicates. To pull up the predicates safely,
+  // other nodes might need to be adjusted:
+  //    - Projection nodes might remove the required columns, so they are simply removed. Since we join with a
+  //      semi/anti join, this does not change the semantics. We run before ColumnPruningRule, which then re-adds
+  //      appropriate pruning projections.
+  //    - Alias nodes are due to the same reasons as projection nodes. Since we can't remove them (that would remove
+  //      the aliased columns) we simply extend them to only add and not remove any columns.
+  //    - Aggregate nodes also remove columns. For now, we only support pulling equals predicates above them. For that
+  //      to work, we extend the aggregate nodes to also group by the columns referenced by all predicates pulled from
+  //      below to above them.
+  //
+  // The only other nodes we pull predicates past are sort and validate nodes, which don't need to be adjusted.
+  auto pull_up_info = prepare_predicate_pull_up(right_tree_root, parameter_mapping);
+  auto is_correlated_parameter = [&](ParameterID id) { return parameter_mapping.find(id) != parameter_mapping.end(); };
+  if (contains_unoptimizable_correlated_parameter_usages(right_tree_root, is_correlated_parameter,
+                                                         pull_up_info.predicate_nodes)) {
+    return _apply_to_inputs(node);
+  }
+
+  // Collect all join predicates into one place
+  for (auto& predicate_info : pull_up_info.predicates) {
+    // Note that we cannot remove the predicate nodes here just yet, because the reformulation might still fail when no
+    // valid primary join predicate is found.
+    join_predicates.emplace_back(std::make_shared<BinaryPredicateExpression>(
+        predicate_info.condition, predicate_info.left_operand, predicate_info.right_operand));
+  }
+
+  // Semi and anti joins are currently only implemented by hash joins. These need an equals comparison as the primary
+  // join predicate.
```

Comment:
And even when having multiple join predicates and the sort-merge join is fully supporting all the remaining modes: using equality predicates as the main join predicate should be a good heuristic that should make sense in the vast majority of cases. So you might want to keep that even with the improved SM join.

Reply:
Shouldn't the optimized version still be faster even if there isn't an equals predicate? It will probably still be slow in most cases, but should still be faster than the unoptimized version, right?